### PR TITLE
Statically ensure that RevCommit instances are never leaked (aka "aliased" in Checker terms)

### PR DESCRIPTION
**Pull request by PawelLipski**
A follow-up for #282

### CODE REVIEW

File: config/checkstyle/tree_walker_module_directives.xml [hunk 1]
```diff
@@ -137,6 +137,12 @@
     <property name="format" value="\bsynchronized\b"/>
     <property name="message" value="Avoid lock-based synchronization, consider Task.Backgroundable and/or GuiUtils.invokeLaterIfNeeded(...) instead"/>
 </module>
+<module name="Regexp">
+    <property name="illegalPattern" value="true"/>
+    <property name="ignoreComments" value="true"/>
+    <property name="format" value="^.*(?&lt;!@Unique |@NonLeaked |import org\.eclipse\.jgit\.revwalk\.)\bRevCommit\b"/>
```
Comment (PawelLipski) [inline config/checkstyle/tree_walker_module_directives.xml:143]: Yes, it's a negative lookbehind :trollface:

File: frontendGraphImpl/src/main/java/com/virtuslab/gitmachete/frontend/graph/impl/repository/RepositoryGraphFactory.java [hunk 2]
```diff
@@ -17,10 +16,8 @@
   private IGitMacheteRepository repository = null;
 
   @Override
-  // Not the most beautiful solution, but let's enforce that this method is only ever called from UI thread
-  // to prevent race conditions on mutable fields.
-  @UIEffect
-  public IRepositoryGraph getRepositoryGraph(IGitMacheteRepository givenRepository, boolean isListingCommits) {
+  @SuppressWarnings("regexp") // to allow `synchronized`
```
Comment (PawelLipski) [inline frontendGraphImpl/src/main/java/com/virtuslab/gitmachete/frontend/graph/impl/repository/RepositoryGraphFactory.java:19]: Unrel

File: config/checker/org.eclipse.jgit.astub [hunk 3]
```diff
@@ -0,0 +1,14 @@
+
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.common.aliasing.qual.NonLeaked;
+import org.checkerframework.common.aliasing.qual.Unique;
+
+package org.eclipse.jgit.revwalk;
+
+class RevWalk {
```
Comment (mkondratek) [inline config/checker/org.eclipse.jgit.astub:8]: This is awesome :1st_place_medal:

File: frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/GitMacheteGraphTable.java [hunk 4]
```diff
@@ -99,9 +99,8 @@ public GitMacheteGraphTable(Project project, IGitRepositorySelectionProvider git
 
     initColumns();
 
-    @SuppressWarnings("guieffect:assignment.type.incompatible")
-    @AlwaysSafe
-    BranchOrCommitCellRenderer branchOrCommitCellRenderer = new BranchOrCommitCellRenderer(/* table */ this, graphCellPainter);
+    @SuppressWarnings("guieffect:assignment.type.incompatible") @AlwaysSafe BranchOrCommitCellRenderer branchOrCommitCellRenderer = new BranchOrCommitCellRenderer(
```
Comment (mkondratek) [inline frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/GitMacheteGraphTable.java:102]: :confused:

Comment (PawelLipski) [reply frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/GitMacheteGraphTable.java:102]: Done, resolved ;)

File: config/checker/org.eclipse.jgit.astub (hunk 3 above)
Comment (PawelLipski) [reply config/checker/org.eclipse.jgit.astub:8]: In the super-ideal world, each value in the program should be immutable; in an almost-super-ideal world where we need to work with external mutable APIs, each value that's of a mutable type should at least be `@Unique` :unicorn:

File: gitCoreJGit/src/main/java/com/virtuslab/gitcore/impl/jgit/GitCoreLocalBranch.java [hunk 5]
```diff
@@ -179,14 +181,17 @@ public String getBranchTypeString(boolean capitalized) {
...
-    RevCommit commit = resolveRevCommit(getPointedCommit().getHash().getHashString());
+    @Unique RevCommit commit = resolveRevCommit(getPointedCommit().getHash().getHashString());
     try {
       walk.markStart(commit);
     } catch (IOException e) {
       throw new GitCoreException(e);
     }
 
-    for (RevCommit currentBranchCommit : walk) {
+    // There's apparently no way for AliasingChecker to work correctly with generics
```
Comment (PawelLipski) [inline gitCoreJGit/src/main/java/com/virtuslab/gitcore/impl/jgit/GitCoreLocalBranch.java:191]: See https://checkerframework.org/manual/#aliasing-unique-restrictions for the reason why

File: gitCoreJGit/src/main/java/com/virtuslab/gitcore/impl/jgit/GitCoreBranch.java [hunk 6]
```diff
@@ -52,36 +53,30 @@ public String getName() {
...
-   */
-  protected RevCommit resolveRevCommit(String revStr) throws GitCoreException {
+  // - - - IMPORTANT NOTE - - -
+  // Bear in mind that RevCommit is a mutable object.
+  // Its internal state (inDegree, flags) changes during a rev walk (among others).
+  // To avoid potential bugs:
+  // - reinstantiate instead of reuse; this method provides a "clean" instance based on the given String revision
+  // - narrow the scope where a RevCommit is available; a use as a field is strongly discouraged.
+  // This comment applies everywhere in the codebase.
+  // Note that both points are kind-of enforced by Checkstyle (every occurrence of "RevCommit" must be preceded with Checker's @Unique annotation),
```
Comment (PawelLipski) [inline gitCoreJGit/src/main/java/com/virtuslab/gitcore/impl/jgit/GitCoreBranch.java:74]: Btw I've just added https://github.com/typetools/checker-framework/issues/3313 (`Feature request: @AlwaysUnique (?) to mark classes whose instances must always be @Unique`)

File: config/checkstyle/tree_walker_module_directives.xml [hunk 7]
```diff
@@ -23,13 +23,15 @@
...
+    <!-- Avoid mutable collections; if really needed, qualify their class names explicitly. -->
+    <property name="illegalClasses" value="java.util.Deque, java.util.List, java.util.Map, java.util.Queue, java.util.Set, java.util.Stack, java.util.Vector"/>
     <!-- Use string interpolation instead (com.antkorwin:better-strings). -->
     <property name="illegalClasses" value="java.text.MessageFormat"/>
     <!-- Use io.vavr.control.Option instead. -->
     <property name="illegalClasses" value="java.util.Optional"/>
     <!-- Use org.checkerframework.checker.nullness.qual.Nullable instead. -->
     <property name="illegalClasses" value="javax.annotation.Nonnull, javax.annotation.Nullable"/>
-    <property name="illegalClasses" value="org.jetbrains.annotations.Nullable, org.jetbrains.annotations.NotNull"/>
+    <property name="illegalClasses" value="org.jetbrains.annotations.NotNull, org.jetbrains.annotations.Nullable"/>
```
Comment (mkondratek) [inline config/checkstyle/tree_walker_module_directives.xml:34]: :xdddd:

File: branchLayoutImpl/src/main/java/com/virtuslab/branchlayout/impl/readwrite/BranchLayoutFileReader.java [hunk 8]
```diff
@@ -171,8 +171,9 @@ private BaseBranchLayoutEntry createEntry(String line, List<BaseBranchLayoutEntr
             "One of branches in branch layout file (${path.toAbsolutePath()}) has incorrect level in relation to its parent branch");
       }
 
-      @SuppressWarnings("index:argument.type.incompatible")
-      Integer upstreamLineIndex = level <= 0 ? -1 : levelToPresentUpstream.get(level - 1);
+      @SuppressWarnings("index:argument.type.incompatible") Integer upstreamLineIndex = level <= 0
```
Comment (micpiotrowski) [inline branchLayoutImpl/src/main/java/com/virtuslab/branchlayout/impl/readwrite/BranchLayoutFileReader.java:174]: Why so weird formatting

File: config/spotless/formatting-rules.xml [hunk 9]
```diff
@@ -108,7 +108,7 @@
         <setting id="org.eclipse.jdt.core.formatter.indent_switchstatements_compare_to_switch" value="true"/>
         <setting id="org.eclipse.jdt.core.formatter.insert_new_line_after_annotation_on_enum_constant" value="insert"/>
         <setting id="org.eclipse.jdt.core.formatter.insert_new_line_after_annotation_on_field" value="insert"/>
-        <setting id="org.eclipse.jdt.core.formatter.insert_new_line_after_annotation_on_local_variable" value="insert"/>
+        <setting id="org.eclipse.jdt.core.formatter.insert_new_line_after_annotation_on_local_variable" value="do not insert"/>
```
Comment (micpiotrowski) [inline config/spotless/formatting-rules.xml:111]: Ok, so we are now using this kind of formatting for local variables?

File: branchLayoutImpl/src/main/java/com/virtuslab/branchlayout/impl/readwrite/BranchLayoutFileReader.java (hunk 8 above)
Comment (PawelLipski) [reply branchLayoutImpl/src/main/java/com/virtuslab/branchlayout/impl/readwrite/BranchLayoutFileReader.java:174]: Line too short :frowning_face:

File: config/spotless/formatting-rules.xml (hunk 9 above)
Comment (PawelLipski) [reply config/spotless/formatting-rules.xml:111]: Yeah, annotation right before the type (see checkstyle changes for why that's needed)

**Pull request merged**